### DISCUSSION TITLE
How to run a socket-activated Nginx container with rootless Podman?

### DISCUSSION THREAD

**Discussion by eriksjolund** [containers/podman#15354] [Q&A] [answered]
How to run a socket-activated Nginx container with rootless Podman?

I'm not sure this is even possible.

A question is whether Nginx supports socket activation.

_"nginx includes an undocumented, internal socket-passing mechanism based on the NGINX environmental variable."_

quote from
https://freedesktop.org/wiki/Software/systemd/DaemonSocketActivation/

See also:
https://trac.nginx.org/nginx/ticket/237

(If I find out the answer to my question, I'll fill it in later)

**Comment by eriksjolund** [comment, marked as answer]
I got a minimal demo working:

https://github.com/eriksjolund/podman-nginx-socket-activation

 The demo shows that it is possible to run a socket-activated nginx container with rootless Podman.

> **Note**
> The _Containerfile_ in the demo builds nginx with many features disabled. Hopefully the demo could be modified to instead use an official nginx container image.

#### Update 22 October 2022:

The container image __docker.io/library/nginx__ now supports socket activation!
(__nginx__ still has  no support for the systemd environment variables `LISTEN_FDS`, `LISTEN_FDNAMES`, `LISTEN_PID` but the environment variable `NGINX` lets you combine a systemd socket unit with nginx)

**Comment by emansom** [comment]
Great stuff! 👍🏻

Have you tried flipping it around? 

As Podman supports systemd as init within the container, I presume it would be possible to self-contain this.
Yielding similar benefits on resource saving, without requiring host modification.

**Reply by eriksjolund** [reply]
> Have you tried flipping it around?

A  good idea! 
Unfortunately systemd has no support for receiving socket-activated sockets from its parent process.

There is a feature request for it:

* https://github.com/systemd/systemd/issues/17764



I believe this is the design you are proposing:


``` mermaid
stateDiagram-v2
    [*] --> systemd: first client connects
    systemd --> podman: socket inherited via fork/exec
    state "OCI runtime" as s2
    podman --> conmon: socket inherited via double fork/exec
    conmon --> s2: socket inherited via fork/exec
    state "systemd running in container" as s3
    s2 --> s3: socket inherited via exec
    s3 --> nginx: socket inherited via fork/exec
```

#### Update 1
I read your comment once more and realise I might have misunderstood. Was your idea to also skip the first systemd and instead have Podman  create the listening socket? I think Podman has no support for that as Podman just inherits the socket-activated from systemd.

#### Update 2
I now noticed you had written some comments below. I'll read the links you've provided.

**Reply by emansom** [reply]
Much simpler: running a `.socket` and a `systemd-socket-proxyd` forwarding to NGINX all inside the container itself.

**Reply by emansom** [reply]
Similar to the instructions noted in [this blog](https://www.redhat.com/en/blog/painless-services-implementing-serverless-rootless-podman-and-systemd), but then with a container running systemd as init that runs the `.socket`, `systemd-socket-proxyd` and NGINX within the container.

**Reply by emansom** [reply]
[Here](https://blog.while-true-do.io/podman-systemd-in-containers/) is a excellent write up on running systemd as init for a container with Podman.

**Reply by emansom** [reply]
(or bypassing the `systemd-socket-proxyd` and utilizing the same `fd=3` hack, both methods should work when all systemd units are contained within the container.)

**Reply by eriksjolund** [reply]
Thanks, now I understand what you mean.  That architecture provides on-demand start of nginx but it misses some other advantages.
When the socket originates from systemd on the host you also get 

* [_Native network performance over the socket-activated socket_](https://github.com/eriksjolund/podman-nginx-socket-activation#native-network-performance-over-the-socket-activated-socket)
* [_Possibility to restrict the network in the container_](https://github.com/eriksjolund/podman-nginx-socket-activation#possibility-to-restrict-the-network-in-the-container)
* [_Possibility to restrict the network in the container, Podman and OCI runtime_](https://github.com/eriksjolund/podman-nginx-socket-activation#possibility-to-restrict-the-network-in-the-container-podman-and-oci-runtime)
* [_The source IP address is preserved_](https://github.com/eriksjolund/podman-nginx-socket-activation#the-source-ip-address-is-preserved)

**Reply by emansom** [reply]
Made a crude diagram to eleborate:

```mermaid
sequenceDiagram
    participant host
    participant container
    participant systemd
    participant proxyd
    participant nginx
    host->>container: TCP :8080
    container->>systemd: .socket unit
    systemd->>proxyd: socket activation
    proxyd->>+nginx: activate NGINX
    nginx-->>-proxyd: exit-idle-time
```

**Reply by emansom** [reply]
Network restriction within the container can be handled by [configuring `nftables` ](https://pablotron.org/articles/nftables-examples/#server-firewall) within the container. That should work fine thanks to netns isolation. For extra paranoid level security, there's [`udica`](https://www.redhat.com/en/blog/generate-selinux-policies-containers-with-udica) 😃

For source IP preservation, there's [`X-Forwarded-For`](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/).

**Reply by eriksjolund** [reply]
For rootless Podman there is also slirp4netns that the traffic passes through. A good thing of letting systemd on the host create the listening socket is that the communication passing through the socket does not need to pass through slirp4netns.
